File: backend/src/app/db/models/audit_log.py
"""Audit log model."""

from __future__ import annotations

from datetime import datetime
from typing import List

import sqlalchemy as sa
from sqlalchemy import Text, text
from sqlalchemy.dialects.postgresql import ARRAY, UUID
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.types import TIMESTAMP

from app.db.base import Base


class AuditLog(Base):
    """Audit log entry for tracking data changes."""

    __tablename__ = "audit_log"

    id: Mapped[str] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=text("gen_random_uuid()"),
    )
    timestamp: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True),
        nullable=False,
        server_default=text("now()"),
    )
    table_name: Mapped[str] = mapped_column(
        Text(),
        nullable=False,
        comment="Name of the table that was modified",
    )
    record_id: Mapped[str] = mapped_column(
        Text(),
        nullable=False,
        comment="Primary key of the modified record",
    )
    action: Mapped[str] = mapped_column(
        Text(),
        nullable=False,
        comment="INSERT, UPDATE, or DELETE",
    )
    user_id: Mapped[str | None] = mapped_column(
        Text(),
        nullable=True,
        comment="Cognito user sub who made the change",
    )
    request_id: Mapped[str | None] = mapped_column(
        Text(),
        nullable=True,
        comment="Lambda request ID for correlation",
    )
    old_values: Mapped[dict | None] = mapped_column(
        sa.dialects.postgresql.JSONB(),
        nullable=True,
        comment="Previous values (for UPDATE/DELETE)",
    )
    new_values: Mapped[dict | None] = mapped_column(
        sa.dialects.postgresql.JSONB(),
        nullable=True,
        comment="New values (for INSERT/UPDATE)",
    )
    changed_fields: Mapped[list[str] | None] = mapped_column(
        ARRAY(Text()),
        nullable=True,
        comment="List of fields that changed (for UPDATE)",
    )
    source: Mapped[str] = mapped_column(
        Text(),
        nullable=False,
        server_default=text("'trigger'"),
        comment="Source of the audit entry: trigger or application",
    )
    ip_address: Mapped[str | None] = mapped_column(
        Text(),
        nullable=True,
        comment="Client IP address if available",
    )
    user_agent: Mapped[str | None] = mapped_column(
        Text(),
        nullable=True,
        comment="Client user agent if available",
    )

File: apps/admin_web/src/types/audit-log.ts
export const AUDITABLE_AUDIT_LOG_TABLES = [
  'assets',
  'asset_access_grants',
  'calendar_manual_blocks',
] as const;

export type AuditLogAction = 'INSERT' | 'UPDATE' | 'DELETE';

export type AuditLogSource = 'trigger' | 'application';

export interface AuditLog {
  id: string;
  table_name: string;
  record_id: string;
  action: AuditLogAction;
  user_id: string | null;
  request_id: string | null;
  old_values: Record<string, unknown> | null;
  new_values: Record<string, unknown> | null;
  changed_fields: string[] | null;
  timestamp: string;
  source: AuditLogSource | string;
  ip_address: string | null;
  user_agent: string | null;
}

export interface AuditLogsFilters {
  table?: string;
  record_id?: string;
  user_id?: string;
  action?: AuditLogAction;
  since?: string;
}
